src/tokenserver/extractors.rs

```diff
@@ -62,44 +109,162 @@ impl FromRequest for TokenserverRequest {
+
+                TokenserverError::internal_error()
+            })?;
+
+            Ok(db)
+        })
+    }
+}
+
 impl FromRequest for TokenData {
     type Config = ();
     type Error = Error;
     type Future = LocalBoxFuture<'static, Result<Self, Self::Error>>;
 
-    fn from_request(req: &HttpRequest, payload: &mut Payload) -> Self::Future {
+    fn from_request(req: &HttpRequest, _payload: &mut Payload) -> Self::Future {
         let req = req.clone();
-        let mut payload = payload.take();
 
         Box::pin(async move {
-            let auth = BearerAuth::from_request(&req, &mut payload).await?;
-            let state = match req.app_data::<Data<ServerState>>() {
-                Some(s) => s,
-                None => {
-                    error!("⚠️ Could not load the app state");
-                    return Err(ValidationErrorKind::FromDetails(
-                        "Internal error".to_owned(),
-                        RequestErrorLocation::Unknown,
-                        Some("state".to_owned()),
-                        None,
-                    )
-                    .into());
+            // The request must have a valid Authorization header
+            let authorization_header = req
+                .headers()
+                .get("Authorization")
+                .ok_or_else(|| TokenserverError::invalid_credentials("Unauthorized"))?
+                .to_str()
+                .map_err(|_| TokenserverError::invalid_credentials("Unauthorized"))?;
+
+            // The request must use Bearer auth
+            if let Some((auth_type, _)) = authorization_header.split_once(" ") {
+                if auth_type.to_ascii_lowercase() != "bearer" {
+                    return Err(TokenserverError::invalid_credentials("Unsupported").into());
                 }
-            };
+            }
+
+            let auth = BearerAuth::extract(&req)
+                .await
+                .map_err(|_| TokenserverError::invalid_credentials("Unsupported"))?;
+            let state = get_server_state(&req)?;
+
             // XXX: tokenserver_state will no longer be an Option once the Tokenserver
             // code is rolled out, so we will eventually be able to remove this unwrap().
             let tokenserver_state = state.tokenserver_state.as_ref().unwrap();
-            tokenserver_state.oauth_verifier.verify_token(auth.token())
+            tokenserver_state
+                .oauth_verifier
+                .verify_token(auth.token())
+                .map_err(Into::into)
         })
     }
 }
 
+#[derive(Debug, PartialEq)]
+struct KeyId {
+    client_state: String,
+    keys_changed_at: i64,
+}
+
+impl FromRequest for KeyId {
+    type Config = ();
+    type Error = Error;
+    type Future = LocalBoxFuture<'static, Result<Self, Self::Error>>;
+
+    fn from_request(req: &HttpRequest, _payload: &mut Payload) -> Self::Future {
+        let req = req.clone();
+
+        Box::pin(async move {
```

Comment:
This logic comes from [here](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/views.py#L244) and [here](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/util.py#L74)